File: ui/desktop_app/lib/src/features/announcements/announcement_service.dart
import '../../core/api_client.dart';
import '../../core/api_date_time.dart';

class Announcement {
  const Announcement({
    required this.id,
    required this.title,
    required this.message,
    required this.publishAtUtc,
    required this.createdAtUtc,
    required this.isPublished,
  });
  final String id, title, message;
  final DateTime publishAtUtc, createdAtUtc;
  final bool isPublished;
  factory Announcement.fromJson(Map<String, dynamic> json) => Announcement(
    id: json['id'] as String,
    title: json['title'] as String? ?? '',
    message: json['message'] as String? ?? '',
    publishAtUtc: parseApiUtc(json['publishAtUtc'] as String),
    createdAtUtc: parseApiUtc(json['createdAtUtc'] as String),
    isPublished: json['isPublished'] == true,
  );
}

class AnnouncementService {
  AnnouncementService(this._api);
  final ApiClient _api;
  Future<List<Announcement>> getAll(String token) async {
    final json = await _api.getJson(
      '/api/admin/announcements?page=1&pageSize=100',
      token: token,
    );
    return (json['items'] as List<dynamic>? ?? const [])
        .map((x) => Announcement.fromJson(x as Map<String, dynamic>))
        .toList();
  }

  Future<void> save(
    String token, {
    String? id,
    required String title,
    required String message,
    required DateTime publishAtUtc,
  }) async {
    final body = {
      'title': title,
      'message': message,
      'publishAtUtc': publishAtUtc.toUtc().toIso8601String(),
    };
    if (id == null) {
      await _api.postJson('/api/admin/announcements', body, token: token);
    } else {
      await _api.putJson('/api/admin/announcements/$id', body, token: token);
    }
  }
}

File: backend/BankingApp.Infrastructure/Services/AnnouncementService.cs
using BankingApp.Application.Announcements;
using BankingApp.Application.AuditLogs;
using BankingApp.Application.Common.Exceptions;
using BankingApp.Application.Common.Pagination;
using BankingApp.Application.Interfaces;
using BankingApp.Domain.Entities;
using BankingApp.Infrastructure.Persistence;
using Microsoft.EntityFrameworkCore;

namespace BankingApp.Infrastructure.Services;

public sealed class AnnouncementService(
    BankingAppDbContext db,
    ICurrentUserService currentUser,
    IAuditLogService audit) : IAnnouncementService
{
    public Task<PagedResult<AnnouncementResponse>> GetPublishedAsync(AnnouncementQuery query, CancellationToken token = default) =>
        QueryAsync(db.SystemAnnouncements.AsNoTracking().Where(x => x.PublishAtUtc <= DateTime.UtcNow), query, token);

    public Task<PagedResult<AnnouncementResponse>> GetAdminAsync(AnnouncementQuery query, CancellationToken token = default)
    {
        EnsureAdmin();
        return QueryAsync(db.SystemAnnouncements.AsNoTracking(), query, token);
    }

    public async Task<AnnouncementResponse> GetAdminByIdAsync(Guid id, CancellationToken token = default)
    {
        EnsureAdmin();
        var value = await db.SystemAnnouncements.AsNoTracking().SingleOrDefaultAsync(x => x.Id == id, token)
            ?? throw new NotFoundException("Announcement was not found.");
        return Map(value, DateTime.UtcNow);
    }

    public async Task<AnnouncementResponse> CreateAsync(AnnouncementWriteRequest request, CancellationToken token = default)
    {
        EnsureAdmin();
        var (title, message, publishAt) = Validate(request);
        var value = new SystemAnnouncement
        {
            Id = Guid.NewGuid(), Title = title, Message = message, PublishAtUtc = publishAt,
            CreatedAtUtc = DateTime.UtcNow, CreatedByAdminId = currentUser.UserId
        };
        db.SystemAnnouncements.Add(value);
        await audit.RecordAsync(new AuditLogRecordRequest
        {
            Action = AuditLogActions.AnnouncementCreated,
            EntityType = AuditEntityTypes.SystemAnnouncement,
            EntityId = value.Id.ToString(),
            Description = $"System announcement '{title}' was created.",
            NewValue = publishAt.ToString("O")
        }, token);
        await db.SaveChangesAsync(token);
        return Map(value, DateTime.UtcNow);
    }

    public async Task<AnnouncementResponse> UpdateAsync(Guid id, AnnouncementWriteRequest request, CancellationToken token = default)
    {
        EnsureAdmin();
        var value = await db.SystemAnnouncements.SingleOrDefaultAsync(x => x.Id == id, token)
            ?? throw new NotFoundException("Announcement was not found.");
        var (title, message, publishAt) = Validate(request);
        var oldPublishAt = value.PublishAtUtc;
        value.Title = title; value.Message = message; value.PublishAtUtc = publishAt;
        await audit.RecordAsync(new AuditLogRecordRequest
        {
            Action = AuditLogActions.AnnouncementUpdated,
            EntityType = AuditEntityTypes.SystemAnnouncement,
            EntityId = value.Id.ToString(),
            Description = $"System announcement '{title}' was updated.",
            OldValue = oldPublishAt.ToString("O"), NewValue = publishAt.ToString("O")
        }, token);
        await db.SaveChangesAsync(token);
        return Map(value, DateTime.UtcNow);
    }

    private static async Task<PagedResult<AnnouncementResponse>> QueryAsync(IQueryable<SystemAnnouncement> source, AnnouncementQuery request, CancellationToken token)
    {
        if (!string.IsNullOrWhiteSpace(request.Search))
        {
            var search = request.Search.Trim();
            source = source.Where(x => x.Title.Contains(search) || x.Message.Contains(search));
        }
        var total = await source.CountAsync(token);
        var now = DateTime.UtcNow;
        var entities = await source.OrderByDescending(x => x.PublishAtUtc)
            .Skip((request.Page - 1) * request.PageSize).Take(request.PageSize).ToListAsync(token);
        return new PagedResult<AnnouncementResponse>
        {
            Items = entities.Select(x => Map(x, now)).ToList(), Page = request.Page,
            PageSize = request.PageSize, TotalCount = total
        };
    }

    private static (string Title, string Message, DateTime PublishAtUtc) Validate(AnnouncementWriteRequest request)
    {
        var title = request.Title.Trim(); var message = request.Message.Trim();
        if (title.Length == 0) throw new BusinessException("Title is required.");
        if (message.Length == 0) throw new BusinessException("Message is required.");
        if (title.Length > 160) throw new BusinessException("Title cannot exceed 160 characters.");
        if (message.Length > 2000) throw new BusinessException("Message cannot exceed 2000 characters.");
        if (request.PublishAtUtc == default) throw new BusinessException("Publish date is required.");
        var utc = request.PublishAtUtc.Kind == DateTimeKind.Utc ? request.PublishAtUtc : request.PublishAtUtc.ToUniversalTime();
        return (title, message, utc);
    }

    private void EnsureAdmin()
    {
        if (!currentUser.IsAdmin) throw new UnauthorizedAccessException("Admin access is required.");
    }

    private static AnnouncementResponse Map(SystemAnnouncement value, DateTime now) => new()
    {
        Id = value.Id, Title = value.Title, Message = value.Message, PublishAtUtc = value.PublishAtUtc,
        CreatedAtUtc = value.CreatedAtUtc, CreatedByAdminId = value.CreatedByAdminId,
        IsPublished = value.PublishAtUtc <= now
    };
}
